[make:listener] Add global command for listener and subscriber

StevenRenaux opened the pull request:
Related to #1301 I also suggest to add a global maker for both, Listener and Subscriber.

I worked with the based class `MakeSubscriber` and updated it to make this solution.

In the terminal you can run `make:event` and a question will be displayed like that to know which event class you want to create.

```
$ php bin/console make:event FooBar kernel.request


 What event class to generate(enter ? to see all types) [Subscriber]:
 > ?

Event Types Class
  * Subscriber
  * Listener

 What event class to generate(enter ? to see all types) [Subscriber]:
 > 
```

And if you write well what you want as `FooBarSubscriber` or `FooBarListener` and the event to listen, your file will be generated as Listener or Subscriber.

```php
namespace App\EventListener;

use Symfony\Component\EventDispatcher\Attribute\AsEventListener;
use Symfony\Component\HttpKernel\Event\RequestEvent;
use Symfony\Component\HttpKernel\KernelEvents;

final class FooBarListener
{
    #[AsEventListener(event: KernelEvents::REQUEST)]
    public function onKernelRequest(RequestEvent $event): void
    {
        // ...
    }
}
```

Jean-Beru commented on src/Resources/config/makers.xml:
```diff
@@ -51,6 +51,11 @@
                 <tag name="maker.command" />
             </service>
 
+            <service id="maker.maker.make_event" class="Symfony\Bundle\MakerBundle\Maker\MakeEvent">
```
I think we have to keep the `maker.maker.make_subscriber` to avoid breaking changes if it a service decorates it.

StevenRenaux replied:
Revert deletion and add deprecation

Jean-Beru commented on tests/Maker/MakeEventTest.php:
```diff
@@ -0,0 +1,202 @@
+
+                $content = file_get_contents($runner->getPath('src/EventSubscriber/FooSubscriber.php'));
+
+                self::assertStringContainsString(
+                    'KernelEvents::REQUEST => \'onKernelRequest\'',
+                    $content
+                );
```
Maybe we can compare generated code with a file in `Tests/Maker/fixtures/`, WDYT ?

StevenRenaux replied:
Updated if you want to check ;) [9c96bbf](https://github.com/symfony/maker-bundle/pull/1366/commits/9c96bbf350b7484cc8650e3f86bacde96c09ec6c)

StevenRenaux commented on src/Maker/MakeEvent.php:
```diff
@@ -23,42 +23,78 @@
+ * @author Steven Renaux <steven.renaux8000@gmail.com>
  */
-final class MakeSubscriber extends AbstractMaker
+final class MakeEvent extends AbstractMaker
```
Fixed with this [commit](https://github.com/symfony/maker-bundle/pull/1366/commits/3c1a24137ab17249d259f4bea5a1fa09c243c167)

StevenRenaux commented on src/Maker/MakeEvent.php:
```diff
@@ -23,42 +23,78 @@
     public static function getCommandName(): string
     {
-        return 'make:subscriber';
+        return 'make:event';
```
I don't know because for me it's also sounds like disturbing to me to write `make:subscriber` to get a listener at the end.

Maybe let see what the others reviewers will think about the naming. It's not written in stone :wink:

Jean-Beru replied:
IMO, a subscriber is no more than a listener which brings its configuration (event => method). `make:event-listener --as-subscriber` could be an alternative.

maelanleborgne replied:
It might be too complex, but here is a suggestion :
With `make:event-listener` as the command name, we could have some logic that checks :
- If there's a `--as-subscriber` or `--as-listener` option. If so, we stick with that.
- Else, try to parse the classname's suffix. If we can determine what type of class the user wants to generate, stick with that.
- Else, ask a question, with `EventListener` as default

weaverryan replied:
Yea, if we call it `make:event-listener`... you can still run it with `make:event` :). But I probably like `make:listener` best. We think of "event listeners" in Symfony, JavaScript, etc. "listener" is a dominate name that has a lot of meaning. 

> If there's a --as-subscriber or --as-listener option. If so, we stick with that.

This is a nice idea, but I don't think we need it. I'm cool with looking at the suffix of that class and, if it's not `Listener` or `Subscriber`, asking them which they want. Or, maybe better, just using a listener, as they are so nice nowadays thanks to `AsEventListener`.

weaverryan commented on src/DependencyInjection/CompilerPass/MakeCommandRegistrationPass.php:
```diff
@@ -63,6 +63,14 @@ public function process(ContainerBuilder $container): void
+            if (method_exists($class, 'getCommandAlias')) {
+                $alias = $class::getCommandAlias();
+                $commandDefinition->addTag('console.command', ['command' => $alias, 'description' => 'Deprecated alias of "make:event"']);
+            }
```
Should/could we leverage the `getCommandAliases()` referenced above since that logic already exists?

squrious replied:
+1, I found it a little bit weird to have methods like `getCommandAliases` whose existence leads to messages for very specific classes.

Maybe something like this?
```php
if (method_exists($class, 'getDeprecatedCommandAliases')) {
    foreach ($class::getDeprecatedCommandAliases() as $alias) {
        $commandDefinition->addTag('console.command', ['command' => $alias, 'description' => \sprintf('Deprecated alias of "%s"', $class::getCommandName())]);
    }
}
```

weaverryan commented on src/Maker/MakeEvent.php:
```diff
@@ -0,0 +1,257 @@
+    public static function getCommandAlias(): string
+    {
+        return 'make:subscriber';
+    }
```
Wait, will this mean that `make:subscriber` will now call THIS maker? If so, let's not do this. Instead, let's keep `make:subscriber` as it is, but deprecate it.

StevenRenaux replied:
I made a deprecation and followed the tricks as it was done for `make:unit-test` and `make:functional-test`.
When `make:subscriber` is called it run my maker and alert the user.

https://github.com/symfony/maker-bundle/blob/ab025cd05b93e9b3955e7416cefede5e87fb9deb/src/Maker/MakeTest.php#L60

weaverryan replied:
Hmm, ok fair enough. We are free to change what these commands due between minor releases. So let's give them the better experience.

weaverryan commented on src/Maker/MakeEvent.php:
```diff
@@ -0,0 +1,257 @@
+    public function configureCommand(Command $command, InputConfiguration $inputConfig): void
+    {
+        $command
+            ->addArgument('name', InputArgument::OPTIONAL, 'Choose a class name for your event subscriber or listener (e.g. <fg=yellow>ExceptionSubscriber</> or <fg=yellow>ExceptionListener</>)')
```
```suggestion
            ->addArgument('name', InputArgument::OPTIONAL, 'Choose a class name for your event subscriber or listener (e.g. <fg=yellow>ExceptionListener</> or <fg=yellow>ExceptionSubscriber</>)')
```

weaverryan reviewed (commented):
This is a cool idea 👍

StevenRenaux commented:
@weaverryan  Command renamed make:listener  with listener as default type. 

To resume:
If Listener or Subscriber was not found in the suffix

```
$ php bin/console make:listener foo


 What event class to generate(enter ? to see all types) [Listener]:
 > 
```

If Listener or Subscriber was  found in the suffix we display the originals events to choose, as the legacy command did.

```
$ php bin/console make:listener FooListener

 Suggested Events:
 * Generator::class ()
 * Symfony\Component\Mailer\Event\MessageEvent (Symfony\Component\Mailer\Event\MessageEvent)
 * Symfony\UX\TwigComponent\Event\PostRenderEvent (Symfony\UX\TwigComponent\Event\PostRenderEvent)
 * Symfony\UX\TwigComponent\Event\PreCreateForRenderEvent (Symfony\UX\TwigComponent\Event\PreCreateForRenderEvent)
/* ...*/

  What event do you want to listen to?:
 > 
```

And if no name was filled, a question is displayed:

```
$ php bin/console make:listener

 Choose a class name for your event listener or subscriber (e.g. ExceptionListener or ExceptionSubscriber):
 > 

```
I didn't really understand your comment, if I was right to notify a deprecation about make:subscriber :
#discussion_r1353072629

If it good for you, maybe we would need to create a new Issue for the next major version to clean all of this deprecation as:

- [MakerExtension](https://github.com/StevenRenaux/maker-bundle/blob/f04136d0b5dfbf2d2cba95e30a1981c887fbf122/src/DependencyInjection/MakerExtension.php#L36)
- [MakerExtension](https://github.com/StevenRenaux/maker-bundle/blob/f04136d0b5dfbf2d2cba95e30a1981c887fbf122/src/DependencyInjection/MakerExtension.php#L58)
- [MakeListener](https://github.com/StevenRenaux/maker-bundle/blob/f04136d0b5dfbf2d2cba95e30a1981c887fbf122/src/Maker/MakeListener.php#L52)
- [MakeListener](https://github.com/StevenRenaux/maker-bundle/blob/f04136d0b5dfbf2d2cba95e30a1981c887fbf122/src/Maker/MakeListener.php#L76)
- [MakeListener](https://github.com/StevenRenaux/maker-bundle/blob/f04136d0b5dfbf2d2cba95e30a1981c887fbf122/src/Maker/MakeListener.php#L244)
- [MakeCommandRegistrationPass](https://github.com/StevenRenaux/maker-bundle/blob/f04136d0b5dfbf2d2cba95e30a1981c887fbf122/src/DependencyInjection/CompilerPass/MakeCommandRegistrationPass.php#L69)

PS: not related to this PR but we also have deprecated code about make:functinal-test  and make:unit-test .

weaverryan commented on src/Maker/MakeListener.php:
```diff
@@ -0,0 +1,257 @@
+            while (null === $classToGenerate) {
+                $defaultEvent = 'Listener';
+                $question = new Question('What event class to generate(enter <comment>?</comment> to see all types)', $defaultEvent);
+                $question->setAutocompleterValues(self::ALL_TYPES);
```
This makes more sense as a choice question: either choose `Listener` or `Susbcriber`.

StevenRenaux replied:
Really better, thanks for the tips :wink:

weaverryan commented on src/Maker/MakeListener.php:
```diff
@@ -0,0 +1,257 @@
+
+            while (null === $classToGenerate) {
+                $defaultEvent = 'Listener';
+                $question = new Question('What event class to generate(enter <comment>?</comment> to see all types)', $defaultEvent);
```
> Do you want to generate an event listener or subscriber?

weaverryan commented:
> I didn't really understand your comment, if I was right to notify a deprecation about make:subscriber:

It looks correct what you did 👍 

> If it good for you, maybe we would need to create a new Issue for the next major version to clean all of this deprecation as:

Yes - though no issue needed, this will naturally happen. I don't know when we might release a new major version, but the main thing we will do when that happens is hunt down all of the deprecated code paths and remove them. Assuming we've done a good job marking this with `@deprecated` or `trigger_deprecation()`, they should be easy to find.

weaverryan commented on src/DependencyInjection/MakerExtension.php:
```diff
@@ -48,6 +52,14 @@ public function load(array $configs, ContainerBuilder $container): void
+        $deprecParamsSubscriber = method_exists(Definition::class, 'getDeprecation') ? ['symfony/maker-bundle', '1.51', self::SUBSCRIBER_MAKER_DEPRECATION_MESSAGE] : [true, self::SUBSCRIBER_MAKER_DEPRECATION_MESSAGE];
+        $container
+            ->getDefinition('maker.maker.make_subscriber')
+            ->setDeprecated(...$deprecParamsSubscriber);
```
Actually, `getDeprecation()` has now existed for a long time and exists in every version that Maker supports. So can you simplify this - and while you're can you simplify the code above also for the test stuff?

weaverryan reviewed (commented):
Don't forget to check the php-cs CI job for some coding standards stuff :)

StevenRenaux commented:
Need help @weaverryan , I don't understand why it failed in the appveyor. It only failed when Listener are generated. But in local or during the other tests in the CI it work.
Example on local, on the left side the file generated and in the right the one compare with. I don't understant why we loose an indention during appveyor test.

![image](https://github.com/symfony/maker-bundle/assets/59167761/153da7e2-0ed6-4cdd-9008-4b3947b2767c)

By the way I updated the deprecated messages and finally removed it from MakerExtension.php and added it in the maker.xml

weaverryan commented on src/Resources/skeleton/event/Listener.tpl.php:
```diff
@@ -0,0 +1,14 @@
+
+<?= $use_statements; ?>
+
+final class <?= $class_name ?>
```
Add a `\n` on the end - see https://github.com/symfony/maker-bundle/blob/main/src/Resources/skeleton/message/Message.tpl.php#L5

If `?>` is the last thing on a line, the PHP sucks up the whitespace before the next char. Not sure why we're seeing this on Windows only, but I bet it'll fix it (maybe this odd behavior is windows-only)

StevenRenaux replied:
It's okay about my generated files in appveyor, thanks @weaverryan :wink: 

The two errors left are about `TemplateLinterTest`

weaverryan commented:
Thank you Steven!
